src/Foundatio.Mediator.Abstractions/Mediator.cs

```diff
@@ -21,34 +25,82 @@ public Mediator(IServiceProvider serviceProvider, MediatorConfiguration? configu
+#if !DISABLE_MEDIATOR_OPENTELEMETRY
+        using var activity = ActivitySource.StartActivity("mediator.invoke");
+        if (activity != null)
+        {
+            var messageType = message.GetType();
+            activity.SetTag("messaging.operation", "invoke");
+            activity.SetTag("messaging.message_type", messageType.FullName);
+        }
+#endif
         var handlerFunc = GetInvokeDelegate(message.GetType());
         handlerFunc(this, message, cancellationToken);
     }
 
     public async ValueTask<TResponse> InvokeAsync<TResponse>(object message, CancellationToken cancellationToken = default)
     {
+#if !DISABLE_MEDIATOR_OPENTELEMETRY
+        using var activity = ActivitySource.StartActivity("mediator.invoke");
+        if (activity != null)
+        {
+            var messageType = message.GetType();
+            activity.SetTag("messaging.operation", "invoke");
+            activity.SetTag("messaging.message_type", messageType.FullName);
+            activity.SetTag("messaging.response_type", typeof(TResponse).FullName);
+        }
+#endif
         var handlerFunc = GetInvokeAsyncResponseDelegate(message.GetType(), typeof(TResponse));
         var result = await handlerFunc(this, message, cancellationToken);
         return (TResponse)result!;
     }
 
     public TResponse Invoke<TResponse>(object message, CancellationToken cancellationToken = default)
     {
+#if !DISABLE_MEDIATOR_OPENTELEMETRY
+        using var activity = ActivitySource.StartActivity("mediator.invoke");
+        if (activity != null)
+        {
+            var messageType = message.GetType();
+            activity.SetTag("messaging.operation", "invoke");
+            activity.SetTag("messaging.message_type", messageType.FullName);
+            activity.SetTag("messaging.response_type", typeof(TResponse).FullName);
+        }
+#endif
         var handlerFunc = GetInvokeResponseDelegate(message.GetType(), typeof(TResponse));
         var result = handlerFunc(this, message, cancellationToken);
         return (TResponse)result!;
     }
 
     public ValueTask PublishAsync(object message, CancellationToken cancellationToken = default)
     {
-        var handlersList = GetAllApplicableHandlers(message).ToList();
-        return _configuration.NotificationPublisher.PublishAsync(this, handlersList, message, cancellationToken);
+        var handlersList = GetAllApplicableHandlers(message);
+#if !DISABLE_MEDIATOR_OPENTELEMETRY
+        using var activity = ActivitySource.StartActivity("mediator.publish");
+        if (activity != null)
+        {
+            var messageType = message.GetType();
+            activity.SetTag("messaging.operation", "publish");
+            activity.SetTag("messaging.message_type", messageType.FullName);
+            activity.SetTag("messaging.handler_count", handlersList.Length);
+        }
+#endif
+        return _configuration.NotificationPublisher.PublishAsync(this, handlersList.ToList(), message, cancellationToken);
```

Comment:
Revert all the changes in the Mediator class. The activities should just be in the strongly typed handler method that is generated from the handler generator.